crates/subspace-farmer/src/plot.rs

```diff
@@ -546,78 +546,77 @@ impl PlotWorker {
-        let range = {
-            let current_piece_count = self.piece_count.load(Ordering::SeqCst);
-            let pieces_to_plot =
-                (self.max_piece_count - current_piece_count).min((pieces.count()) as u64);
-
-            let start_offset: PieceOffset = current_piece_count;
+        let current_piece_count = self.piece_count.load(Ordering::SeqCst);
+        let pieces_left_until_full_plot =
+            (self.max_piece_count - current_piece_count).min(pieces.count() as u64);
+
+        // Split pieces and indexes in those that can be appended to the end of plot (thus written
+        // sequentially) and those that need to be checked individually and plotted one by one in
+        // place of old pieces
+        let (sequential_pieces, _) =
+            pieces.split_at(pieces_left_until_full_plot as usize * PIECE_SIZE);
+        // Iterator is more convenient for random pieces, otherwise we could take it from above
+        let random_pieces = pieces
+            .as_pieces()
+            .skip(pieces_left_until_full_plot as usize);
+        let (sequential_piece_indexes, random_piece_indexes) =
+            piece_indexes.split_at(pieces_left_until_full_plot as usize);
 
+        // Process sequential pieces
+        {
             self.plot
-                .seek(SeekFrom::Start(start_offset * PIECE_SIZE as u64))?;
-            self.plot
-                .write_all(&pieces[..pieces_to_plot as usize * PIECE_SIZE])?;
-
-            let piece_indexes: Vec<PieceIndex> = piece_indexes
-                .iter()
-                .take(pieces_to_plot as usize)
-                .copied()
-                .collect();
+                .seek(SeekFrom::Start(current_piece_count * PIECE_SIZE as u64))?;
+            self.plot.write_all(sequential_pieces)?;
 
-            for (index, offset) in piece_indexes.iter().copied().zip(start_offset..) {
-                self.piece_index_hash_to_offset_db.put(index, offset)?;
-                self.put_piece_index(offset, index)?;
+            for (piece_offset, &piece_index) in
+                (current_piece_count..).zip(sequential_piece_indexes)
+            {
+                self.piece_index_hash_to_offset_db
+                    .put(piece_index, piece_offset)?;
+                self.put_piece_index(piece_offset, piece_index)?;
             }
 
-            self.piece_count.fetch_add(pieces_to_plot, Ordering::AcqRel);
-
-            start_offset..start_offset + pieces_to_plot
-        };
+            self.piece_count
+                .fetch_add(pieces_left_until_full_plot, Ordering::AcqRel);
+        }
 
-        // Overwrite pieces
         let mut offsets = Vec::<PieceOffset>::with_capacity(pieces.count());
-        offsets.extend(range.clone());
-        let mut old_pieces = Vec::with_capacity(pieces.count());
+        offsets.extend((current_piece_count..).take(pieces_left_until_full_plot as usize));
+        let mut evicted_pieces = Vec::with_capacity(pieces.count());
```

Comment:
We might do big and useless allocation here, if it is just a startup. May be we can add a check `if random_pieces.is_empty()` beforehand. Just in case

Reply:
Didn't notice last commit